File: spec/01_arrays_and_strings/1_1_is_unique.rb
# str.size = n
# alphabet(str) = a (size of alphabet)

# A: time = O(n), mem = O(a)
def all_uniq_a?(str)
  h = {}

  str.each_char do |c|
    if h[c]
      return false
    else
      h[c] = true
    end
  end

  true
end

# C: time = O(n*log(n)), mem = O(1)
def all_uniq_c?(str)
  str = sort!(str)

  (str.size - 1).times do |i|
    if str[i] == str[i + 1]
      return false
    end
  end

  true
end

# E: time = O(n), mem = O(1)
# Special case if alphabet <= 64 (for 64 bits systems), and alphabet is continuous
def all_uniq_e?(str)
  bits = 0

  str.each_char do |c|
    bit = 2 ** (c.ord - 'a'.ord)

    if bits & bit == 1
      return false
    else
      bits |= bit
    end
  end

  true
end

RSpec.describe 'all_uniq?' do
  [:a, :c, :e].each do |implementation|
    describe "#{implementation} case" do
      subject { send("all_uniq_#{implementation}?", str) }

      context 'all uniq' do
        let(:str) { 'abcdefg' }

        it { is_expected.to eq true }
      end

      context 'not all uniq' do
        let(:str) { 'abcdefga' }

        it { is_expected.to eq false }
      end
    end
  end
end

File: go/01_arrays_and_strings/1_1_is_unique.go
package main

// str.size = n
// alphabet(str) = a (size of alphabet)

import (
	"sort"
)

// A: time = O(n), mem = O(a)
func AllUniqA(str string) bool {
	h := make(map[rune]bool)

	for _, c := range str {
		if h[c] {
			return false
		} else {
			h[c] = true
		}
	}

	return true
}

// C: time = O(n*log(n)), mem = O(1)
func AllUniqC(str string) bool {
	str = SortStringByCharacter(str)
	runes := StringToRuneSlice(str)

	for i := 0; i < len(runes) - 1; i++ {
	  if str[i] == str[i + 1] {
	    return false
	  }
	}

	return true
}

func StringToRuneSlice(s string) []rune {
  var r []rune

  for _, runeValue := range s {
    r = append(r, runeValue)
  }

  return r
}

func SortStringByCharacter(s string) string {
  r := StringToRuneSlice(s)

  sort.Slice(r, func(i, j int) bool {
    return r[i] < r[j]
  })

  return string(r)
}

// E: time = O(n), mem = O(1)
// Special case if alphabet <= 64 (for 64 bits systems), and alphabet is continuous
func AllUniqE(str string) bool {
  bits := 0

  for _, c := range str {
    bit := 1 << uint(c - 'a')

    if bits & bit == 1 {
      return false
    } else {
      bits |= bit
    }
  }

  return true
}
